Controller/Giftcard/AddGiftCard.php

```diff
@@ -0,0 +1,225 @@
+
+            // Validate quote
+            if (!$quote || !is_object($quote)) {
+                return $this->jsonResponse(false, __('We couldn\'t retrieve your cart details. Please try again.'));
+            }
+
+            if ($isCartPage && !$quote->isVirtual()) {
+                $this->updateShippingMethod($quote, $selectedShippingMethod);
+            }
+
+            if ($this->isCartTotalZero($quote)) {
+                return $this->jsonResponse(false, __('Your cart total is zero, so there\'s no need to apply a Gift Card code.'));
+            }
+
+            $apiResponse = $this->giftcardHelper->checkGiftCardBalanceStatus($giftCardCode);
+
+            return $this->handleApiResponse($apiResponse, $giftCardCode, $quote);
+
+        } catch (LocalizedException $e) {
+            return $this->jsonResponse(false, $e->getMessage());
+        }
+    }
+
+    /**
+     * Update the shipping method of the quote
+     *
+     * @param mixed $quote
+     * @param mixed $selectedShippingMethod
+     * @return void
+     */
+    private function updateShippingMethod($quote, $selectedShippingMethod)
+    {
+        $shippingMethod = $quote->getShippingAddress()->getShippingMethod();
+        $shippingAmount = $quote->getShippingAddress()->getShippingAmount();
+
+        if ($selectedShippingMethod &&
+            ($shippingMethod !== $selectedShippingMethod || $shippingAmount == 0)) {
+            $this->giftcardHelper->setShippingMethodForce(
+                $quote,
+                $selectedShippingMethod
+            );
+        }
+    }
+
+    /**
+     * Check if the cart total is zero
+     *
+     * @param mixed $quote
+     * @return bool
+     */
+    private function isCartTotalZero($quote)
+    {
+        return (float)$quote->getGrandTotal() === 0.0 && $quote->getShippingAddress()->getShippingAmount() == 0;
+    }
+
+    /**
+     * Handle the API response for gift card validation.
+     *
+     * @param mixed $apiResponse
+     * @param mixed $giftCardCode
+     * @param mixed $quote
+     * @return mixed
+     */
+    private function handleApiResponse($apiResponse, $giftCardCode, $quote)
+    {
+        $apiErrorMessage = $apiResponse['xError'] ?? null;
+        $status = $apiResponse['xStatus'] ?? null;
+        $activationStatus = $apiResponse['xActivationStatus'] ?? null;
+        $remainingBalance = $apiResponse['xRemainingBalance'] ?? null;
+
+        if ($status === "Approved" && $activationStatus === "Active" && $remainingBalance > 0) {
+            return $this->applyGiftCard($apiResponse, $giftCardCode, $quote);
+        }
+
+        if ($activationStatus === 'Inactive') {
+            return $this->jsonResponse(false, __('Your gift card account is inactive'));
+        }
+
+        if ((float)$remainingBalance === 0.0) {
+            return $this->jsonResponse(false, __('Your gift card balance is zero. Please use another card number or credit your balance.'));
```

Comment:
'Your gift card balance is zero. Please use another card.'

Reply:
@ybroyde , Updated error message.